src/search/BracketsSearchCursor.js

```diff
@@ -0,0 +1,823 @@
+     *
+     * @param {Array<number>} lineCharacterCountIndexArray See '_createLineCharacterCountIndex'
+     * @param {String}        documentText                 Text to scan
+     * @param {RegExp}        regex                        Regular expression used to search
+     * @param {{from: {line: number, ch: number}, to: {line: number, ch: number}}} range Area to scan for matches
+     * @param {function({line: number, ch: number}, {line: number, ch: number}, Array )} fnEachMatch Function is called with start position, end position and Regex match Array
+     */
+    function _scanDocumentUsingRegularExpression(documentIndex, documentText, regex, range, fnEachMatch) {
+        var matchArray;
+        var startRangeIndex = range !== undefined ? _indexFromPos(documentIndex, range.from) : 0;
+        var endRangeIndex = range !== undefined ? _indexFromPos(documentIndex, range.to) : documentIndex[documentIndex.length - 1];
+
+        regex.lastIndex = startRangeIndex;
+        var lastMatchedLine = 0;
+        while ((matchArray = regex.exec(documentText)) !== null) {
+            var startPosition = _createPosFromIndex(documentIndex, lastMatchedLine, matchArray.index);
+            var endPosition = _createPosFromIndex(documentIndex, startPosition.line, regex.lastIndex);
+            lastMatchedLine = endPosition.line;
+
+            if (regex.lastIndex <= endRangeIndex) {
+                fnEachMatch(startPosition, endPosition, matchArray);
+            } else {
+                break;
+            }
+            // This is to stop infinite loop.  Some regular expressions can return 0 length match
+            // which will not advance the lastindex property.  Ex ".*"
+            if (matchArray.index === regex.lastIndex) {
+                regex.lastIndex++;
+            }
+        }
+    }
+
+
+    /**
+     * Returns an object that indicates the beginning and end of a match from the search
+     * @param {Array<number>} docLineIndex See '_createLineCharacterCountIndex'
+     * @param {number}        indexStart   Start location using index
+     * @param {number}        indexEnd     End location using index
+     * @param {number}        startLine    Starting line to search for line locations
+     *
+     */
+    function _createSearchResult(docLineIndex, indexStart, indexEnd, startLine) {
+        if (typeof startLine === 'undefined') {
+            startLine = 0;
+        }
+
+        var fromPos = _createPosFromIndex(docLineIndex, startLine, indexStart);
+        var toPos   = _createPosFromIndex(docLineIndex, fromPos.line, indexEnd);
+
+        return {from: fromPos, to: toPos};
+    }
+
+    /**
+     * Comparison function for binary search of index positions within document.
+     * @param {number} matchIndex First match index for compare
+     * @param {number} posIndex   Second match index for compare
+     * @returns {number} Result of comparison
+     */
+    function _compareMatchResultToPos(matchIndex, posIndex) {
```

Comment:
Names `matchIndex` and `posIndex` are irrelevant. So if the long name of the function. Since you separate and declare comparator, it should be just `_compareOffsets()`.